### PR TITLE
feat: streams based `Buffer`

### PR DESCRIPTION
**Pull request by crowlKats**
Equivalent of `std/io/buffer.ts`'s `Buffer` but with streams

### CODE REVIEW

File: streams/buffer.ts [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+  /** The read only capacity of the buffer's underlying byte slice, that is,
+   * the total space allocated for the buffer's data. */
+  get capacity(): number {
+    return this.#buf.buffer.byteLength;
+  }
+
+  /** Discards all but the first `n` unread bytes from the buffer but
+   * continues to use the same allocated storage. It throws if `n` is
+   * negative or greater than the length of the buffer. */
+  truncate(n: number): void {
```
Comment (kt3k) [inline streams/buffer.ts:92]: truncate doesn't seem tested?

Comment (crowlKats) [reply streams/buffer.ts:92]: Done

**Comment by bartlomieju** [comment]
What's the use case for this? Why create a new class with the same name instead of adding this functionality to existing `Buffer`?

**Comment by crowlKats** [comment]
@bartlomieju well, we want various Deno.Reader/Writer based APIs to be written/ported to streams. These should be in their own file so there in no dependency on the aforementioned APIs

**Review by bartlomieju — APPROVED** [review]
LGTM

**Comment by jsejcksn** [comment]
> What's the use case for this? Why create a new class with the same name instead of adding this functionality to existing `Buffer`?

@crowlKats Does this also allow for browser compatibility?

**Comment by crowlKats** [comment]
@jsejcksn nope, once again, typescript is behind the times with outdated typings

**Comment by jsejcksn** [comment]
> @jsejcksn nope, once again, typescript is behind the times with outdated typings

@crowlKats There's a tracking issue for compat issues caused by outdated native TS types [here](https://github.com/denoland/deno_std/issues/1973). Will you add any candidate modules there?

**Comment by crowlKats** [comment]
@jsejcksn done

**Pull request merged**